feat: get variation details by variation type

duyhungtnn opened the pull request:
Related https://github.com/bucketeer-io/bucketeer/issues/886
### Changes
- Added a new data class `BKTEvaluationDetail<T>` to support get variation details by variation type.
- Added a new data class `BKTValue` to reference a JSON value
- Introduce new interface to get variation details by variation type
```kotlin
fun intVariationDetails(
  featureId: String,
  defaultValue: Int,
): BKTEvaluationDetails<Int>

fun doubleVariationDetails(
  featureId: String,
  defaultValue: Double,
): BKTEvaluationDetails<Double>

fun boolVariationDetails(
  featureId: String,
  defaultValue: Boolean,
): BKTEvaluationDetails<Boolean>

fun stringVariationDetails(
  featureId: String,
  defaultValue: String,
): BKTEvaluationDetails<String>

fun objectVariationDetails(
  featureId: String,
  defaultValue: BKTValue,
): BKTEvaluationDetails<BKTValue>

fun objectVariation(
  featureId: String,
  defaultValue: BKTValue,
): BKTValue
```
- The class `BKTEvaluation`, the method `getEvaluationDetails` & `jsonVariation` will be deprecated after this PR.

cre8ivejp commented on bucketeer/src/androidTest/kotlin/io/bucketeer/sdk/android/e2e/BKTClientVariationTest.kt:
```diff
@@ -82,6 +85,21 @@ class BKTClientVariationTest {
+        variationId = "36a53a17-60b4-4a99-a54a-7fcbf21f7c8c",
+        variationName = "variation 1",
+        variationValue = "value-1",
+        reason = BKTEvaluationDetail.Reason.DEFAULT,
```
For unit tests, it would be safer if we had tests to check the default value and the reason, when the feature flag doesn't exist in the cache.
No need to add the same test for e2e tests.

duyhungtnn replied:
I added tests for it 
https://github.com/bucketeer-io/android-client-sdk/pull/180/files#diff-e3063ec01decd033a3d58ff6938582446f6ab91798ce7f7418349e16e98095d0R10

and

https://github.com/bucketeer-io/android-client-sdk/pull/180/files#diff-e3063ec01decd033a3d58ff6938582446f6ab91798ce7f7418349e16e98095d0R10

cre8ivejp commented on bucketeer/src/main/kotlin/io/bucketeer/sdk/android/BKTEvaluationDetail.kt:
```diff
@@ -0,0 +1,84 @@
+    ;
+
+    companion object {
+      fun from(value: String): Reason = entries.firstOrNull { it.name == value } ?: DEFAULT
```
The `CLIENT` means that the SDK returned the default value to the app.
I think it's better if we return the `CLIENT` when we don't find the reason.

```suggestion
      fun from(value: String): Reason = entries.firstOrNull { it.name == value } ?: CLIENT
```

cre8ivejp commented on bucketeer/src/main/kotlin/io/bucketeer/sdk/android/BKTEvaluationDetail.kt:
```diff
@@ -0,0 +1,84 @@
+        variationId = "",
+        variationName = "",
+        variationValue = defaultValue,
+        reason = BKTEvaluationDetail.Reason.DEFAULT,
```
As mentioned [here](https://github.com/bucketeer-io/android-client-sdk/pull/180/files#r1636188652), it must return `CLIENT` when the default value is returned.

```suggestion
        reason = BKTEvaluationDetail.Reason.CLIENT,
```

duyhungtnn replied:
updated in https://github.com/bucketeer-io/android-client-sdk/pull/180/commits/6ce893f005267fbbbfcb19fea7e7e53bd6b0cad2

cre8ivejp commented on bucketeer/src/main/kotlin/io/bucketeer/sdk/android/BKTEvaluationDetail.kt:
```diff
@@ -0,0 +1,84 @@
+    ): BKTEvaluationDetail<T> {
+      return BKTEvaluationDetail(
+        id = "",
+        featureId = "",
```
We must set the `featureId` so the app can know which flag the result is coming from.
Please, also update the tests.

duyhungtnn replied:
I updated it in https://github.com/bucketeer-io/android-client-sdk/pull/180/commits/cfbd5ac6c681782ad6566358f8f3fd42487421e4

cre8ivejp commented on bucketeer/src/main/kotlin/io/bucketeer/sdk/android/BKTEvaluationDetail.kt:
```diff
@@ -0,0 +1,85 @@
+import org.json.JSONObject
+
+data class BKTEvaluationDetail<T>(
+  var id: String,
```
I just found out that we won't need this field to support the OpenFeature SDK.

```suggestion
```

duyhungtnn replied:
updated in https://github.com/bucketeer-io/android-client-sdk/pull/180/commits/bbb4bcede0e56f8c44a568d11cbc50c9bd57472b

cre8ivejp commented:
@duyhungtnn, please fix the conflict.

duyhungtnn commented:
@cre8ivejp I fixed the conflict.

cre8ivejp commented:
@duyhungtnn, sorry for asking for these changes so late.
Can you rename the interfaces as follows? The idea is to keep the same name as the non-details interfaces, too.

E.g.

```kt
intVariation("id", 0)
intVariationDetails("id", 0)
```

```kt
fun intVariationDetails(
  featureId: String,
  defaultValue: Int,
): BKTEvaluationDetail<Int>

fun doubleVariationDetails(
  featureId: String,
  defaultValue: Double,
): BKTEvaluationDetail<Double>

fun boolVariationDetails(
  featureId: String,
  defaultValue: Boolean,
): BKTEvaluationDetail<Boolean>

fun stringVariationDetails(
  featureId: String,
  defaultValue: String,
): BKTEvaluationDetail<String>

fun jsonVariationDetails(
  featureId: String,
  defaultValue: JSONObject,
): BKTEvaluationDetail<JSONObject>
```

cre8ivejp reviewed (approved):
Thank you!

duyhungtnn commented:
hi @cre8ivejp
Please help me to take a look at this PR.
### Changes
- Added a new data class BKTValue to reference a JSON node value
- Added a new method 
 ```kotlin
 fun objectVariationDetails(
  featureId: String,
  defaultValue: BKTValue,
): BKTEvaluationDetails<BKTValue>
```
- Removed the method `jsonVariationDetail`
- Deprecated the method `jsonVariation`

duyhungtnn commented:
### New changes 
- Add support to return intValue even when the underlying value is a double https://github.com/bucketeer-io/android-client-sdk/pull/180/commits/b8c7ab9555485a668d94f1f0e7f5c6a701c8413e

duyhungtnn commented:
### New changes 
[keep the detail interface right after the non-detail interface](https://github.com/bucketeer-io/android-client-sdk/pull/180/commits/9be51b7eafc690296740973c3352ce2fff097ab2)

cre8ivejp commented on bucketeer/src/main/kotlin/io/bucketeer/sdk/android/internal/evaluation/EvaluationExt.kt:
```diff
@@ -1,62 +1,94 @@
+      String::class ->
+        value
+      Int::class -> value.toDoubleOrNull()?.toInt()
+      Long::class -> value.toLongOrNull()
+      Float::class -> value.toFloatOrNull()
```
Just confirming.
Long and Float won't be used, right?

duyhungtnn replied:
yes, we no longer use Long or Float.

duyhungtnn commented on bucketeer/src/main/kotlin/io/bucketeer/sdk/android/BKTClientImpl.kt:
```diff
@@ -107,18 +126,18 @@ internal class BKTClientImpl(
-  }
 
+  @Deprecated("evaluationDetails is deprecated")
+  @Suppress("DEPRECATION")
```
updated https://github.com/bucketeer-io/android-client-sdk/pull/180/commits/fbc015d765cdc2d874e4d729f070b2ea3adeb698

cre8ivejp reviewed (approved):
Thank you!
